### DISCUSSION TITLE
foreign archive handling

### DISCUSSION THREAD

**Discussion by progman1** [ocaml/dune#15150] [Q&A]
I have a (foreign_archive .. ) for which there are .so and .a archives.
when an .exe is built ldd showed it to be dynamically linked against the .so.
some time later after meddling with dune stanzas BUT returning to config for above, 
the .exe built is always resolved and so appears to have incorporated the .a archive.
and now I cannot get the dynamic linking behaviour back.

I am 99.9% certain I am seeing different behaviour from identical configuration!
perhaps dune is holding on to some state?
so rather than hand off a bug report for something I cannot reproduce may I ask what decisions
dune makes around whether to use .a's or .so's ?

**Comment by ghost** [comment]
For native binaries, the `.a` should get linked.

**Comment by progman1** [comment]
and there are no circumstances whatsoever under which a .so is taken?

**Comment by progman1** [comment]
.

**Comment by nojb** [comment]
In order to investigate an issue we would need precise instruction to reproduce the problem; otherwise there is not much we can do, unfortunately.

**Comment by progman1** [comment]
I did ask only for information around dune choices.
as it turned out the issue was with the gnu linker.
although at the dune level foreign_archive in combo with a .exe target forces a .a target, if the linker used by ocaml 
subsequently sees a .so it uses it over the .a.

this is useful as the .so can have all dependent libraries' paths in place and ocaml need not be aware of them (correction: LD_LIBRARY_PATH may need setting up), while
linking against a .a forces taking account in the dune config of those dependent libraries' paths, which seems unfortunate.
also a smaller binary is generated in less time.

while ocamlopt can take .so cmdline arguments I don't know a way to direct dune to use shared libs exclusively.

**Comment by progman1** [comment]
trying to close issue....
...and just stays around.

**Comment by ghost** [comment]
Eventually it would be nice for Dune to support building executables using shared libraries. Including OCaml libraries as well, there is no need to limit ourselves to C libraries. Though that's a fair amount of work, especially if we include teaching Dune how to run such binaries.
